minicbor-derive/src/attrs/idx.rs

```diff
@@ -24,12 +24,24 @@ impl Idx {
     }
 
     /// Get the numeric index value.
-    pub fn val(self) -> u32 {
+    pub fn val(self) -> i32 {
         match self {
             Idx::N(i) => i,
             Idx::B(i) => i
         }
     }
+
+    /// Get a property of the index that is sorted as needed to produce the encoded items:
+    /// They are sorted in their deterministic encoding.
+    ///
+    /// This produces the correct sequence both for CBOR arrays (where all items are emitted in
+    /// ascending order) and for CBOR maps (where keys need to be sorted in bytewise lexicographic
+    /// ordering).
+    ///
+    /// This is used for pre-sorting items for serialization.
+    pub fn val_for_sorting(self) -> impl Ord {
+        (self.val() < 0, self.val().unsigned_abs())
```

Comment:
I do not think this is actually producing a bytewise lexicographic order as it would sort `[-1, -10, -100]` as `[-1, -10, -100]`, however the correct order would be `[-100, -10, -1]` (`[9c, f6, ff]`). The correct approach would be to sort by `i32::to_be_bytes`. I suggest you remove this method and in `Fields` you do `fields.sort_unstable_by_key(|f| f.index.val().to_be_bytes());`.